pkg/crcontrollers/csibmnode/controller.go

```diff
@@ -0,0 +1,328 @@
+func (bmc *Controller) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ll := bmc.log.WithFields(logrus.Fields{
+		"method": "Reconcile",
+		"name":   req.Name,
+	})
+
+	var err error
+	// if name in request doesn't start with namePrefix controller tries to read k8s node object at first
+	// however if it get NotFound error it tries to read CSIBMNode object as well
+	if !strings.HasPrefix(req.Name, namePrefix) {
+		k8sNode := new(coreV1.Node)
+		err = bmc.k8sClient.ReadCR(context.Background(), req.Name, k8sNode)
+		switch {
+		case err == nil:
+			ll.Infof("Reconcile k8s node %s", k8sNode.Name)
+			return bmc.reconcileForK8sNode(k8sNode)
+		case !k8sError.IsNotFound(err):
+			ll.Errorf("Unable to read node object: %v", err)
+			return ctrl.Result{Requeue: true}, err
+		}
+	}
+
+	// try to read CSIBMNode
+	bmNode := new(nodecrd.CSIBMNode)
+	err = bmc.k8sClient.ReadCR(context.Background(), req.Name, bmNode)
+	switch {
+	case err == nil:
+		ll.Infof("Reconcile CSIBMNode %s", bmNode.Name)
+		return bmc.reconcileForCSIBMNode(bmNode)
+	case !k8sError.IsNotFound(err):
+		ll.Errorf("Unable to read CSIBMNode object: %v", err)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	ll.Warnf("unable to detect for which object (%s) that reconcile is. The object may have been deleted", req.String())
+	return ctrl.Result{}, nil
+}
+
+func (bmc *Controller) reconcileForK8sNode(k8sNode *coreV1.Node) (ctrl.Result, error) {
+	ll := bmc.log.WithFields(logrus.Fields{
+		"method": "reconcileForK8sNode",
+		"name":   k8sNode.Name,
+	})
+
+	if len(k8sNode.Status.Addresses) == 0 {
+		err := errors.New("addresses are missing for current k8s node instance")
+		ll.Error(err)
+		return ctrl.Result{Requeue: false}, err
+	}
+
+	// get corresponding CSIBMNode CR name from cache
+	var bmNode = &nodecrd.CSIBMNode{}
+	if bmNodeName, ok := bmc.cache.getCSIBMNodeName(k8sNode.Name); ok {
+		if err := bmc.k8sClient.ReadCR(context.Background(), bmNodeName, bmNode); err != nil {
+			ll.Errorf("Unable to read CSIBMNode %s: %v", bmNodeName, err)
+			return ctrl.Result{Requeue: true}, err
+		}
+		return bmc.updateAnnotation(k8sNode, bmNode.Spec.UUID)
+	}
+
+	// search corresponding CSIBMNode CR name in k8s API
+	bmNodeCRs := new(nodecrd.CSIBMNodeList)
+	if err := bmc.k8sClient.ReadList(context.Background(), bmNodeCRs); err != nil {
+		ll.Errorf("Unable to read CSIBMNode CRs list: %v", err)
+		return ctrl.Result{Requeue: true}, err
+	}
+
+	matchedCRs := make([]string, 0)
+	for i := range bmNodeCRs.Items {
+		matchedAddresses := bmc.matchedAddressesCount(&bmNodeCRs.Items[i], k8sNode)
+		if len(bmNodeCRs.Items[i].Spec.Addresses) > 0 && matchedAddresses == len(bmNodeCRs.Items[i].Spec.Addresses) {
+			bmNode = &bmNodeCRs.Items[i]
+			matchedCRs = append(matchedCRs, bmNode.Name)
+			continue
+		}
+		if matchedAddresses > 0 {
+			ll.Warnf("There is CSIBMNode %s that partially match k8s node %s. CSIBMNode.Spec: %v, k8s node addresses: %v",
```

Comment:
I think this log message can be "upgraded" to Error level, because we have inconsistent state and can't handle this situation. +short instruction in log message can be useful, for example, "CSIBMNode Spec should be edited to match exactly one kubernetes node"

Reply:
done